### PR TITLE
GODRIVER-2197 Defer Implicit Session Checkout

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-2197

The expectation of the ticket is to only check out an implicit session after we check out a connection. To resolve this for the Go driver, I am suggesting we leave everything as-is in the collection CRUD methods concerning `session.NewClientSession`. Within `session.NewClientSession`, I suggest we defer the selection of an implicit session to be set post-function-runtime by the secondary API `(*session.Client).SetServer`. In the case of the collection's CRUD methods, we call `(*session.Client).SetServer` after a connection is checked out in the operation executor.

Here is a sequence diagram for the proposal:
Implicit Session Sequence Diagram (3)

### CODE REVIEW

File: mongo/collection.go [hunk 1]
```diff
@@ -1818,7 +1818,8 @@ func (coll *Collection) drop(ctx context.Context) error {
 	sess := sessionFromContext(ctx)
 	if sess == nil && coll.client.sessionPool != nil {
 		var err error
-		sess, err = session.NewClientSession(coll.client.sessionPool, coll.client.id, session.Implicit)
+		sess, err = session.NewClientSessionSetImplicit(coll.client.sessionPool, coll.client.id,
+			session.Implicit)
```
Comment (prestonvasquez) [inline mongo/collection.go:1822]: I don't think we want to do this "implicit session only after connection" thing for dropping collections. It causes some very strange behavior such as

```
InsertMany error for collection test: bulk write exception: write errors: [E11000 duplicate key error collection: transaction-tests.test index: _id_ dup key: { _id: 0 }]
```

File: mongo/integration/sessions_test.go [hunk 2]
```diff
@@ -358,6 +360,115 @@ func TestSessions(t *testing.T) {
...
+			if len(set) == 1 {
+				maintainedOneSession = true
+			}
+			if len(set) < len(ops) {
+				limitedSessionUse = true
+			}
+		}
+
+		oneSessMsg := "expected one session accross all %v operations for at least 1/%v retries"
+		assert.True(mt, maintainedOneSession, oneSessMsg, len(ops), retrycount)
```
Comment (prestonvasquez) [inline mongo/integration/sessions_test.go:466]: @benjirewis @matthewdale @Neilshweky This is an important thing to note:

This assertion _always_ fails on evergreen for sharded clusters, and only sharded clusters. I can get it to create locally, but only if I add delays in the `(*session.Client).EndSession` method (`time.Sleep(3 *time.Second)`, for instance).  [This is a `MUST` requirement of the spec](https://github.com/mongodb/specifications/blob/master/source/sessions/tests/README.rst#implicit-sessions-only-allocate-their-server-session-after-a-successful-connection-checkout:~:text=Drivers%20MUST%20assert%20that%20exactly%20one%20session%20is%20used%20for%20all%20operations%20at%20least%20once%20across%20the%20retries%20of%20this%20test.), but the way the sequence is currently designed is not thread safe -- a race condition can occur when you end a session after ending a connection.

TL;DR: race conditions occur by virtue of the sequence of events.

How do we want to approach this? I think if we don't want to relax the conditions, we may have to create some kind of semaphore/mutex for ending sessions.

File: mongo/integration/sessions_test.go [hunk 3]
```diff
@@ -358,6 +360,121 @@ func TestSessions(t *testing.T) {
...
+					return name, err
+				}
+				return name, cursor.All(ctx, &bson.A{})
+			},
+		}
+
+		// maintainedOneSession asserts that exactly one session is used for all operations at least once
+		// across the retries of this test.
+		var maintainedOneSession bool
+		var minimumSessionCount int
```
Comment (benjirewis) [inline mongo/integration/sessions_test.go:426]: Maybe a comment for the purpose of `minimumSessionCount`, too?

File: mongo/integration/sessions_test.go [hunk 4]
```diff
@@ -358,6 +360,121 @@ func TestSessions(t *testing.T) {
...
+
+			setSize := len(set)
+			if setSize == 1 {
+				maintainedOneSession = true
+			} else if setSize < minimumSessionCount || minimumSessionCount == 0 {
+				// record the minimum number of sessions we used over all retries.
+				minimumSessionCount = setSize
+			}
+
+			if len(set) < len(ops) {
```
Comment (benjirewis) [inline mongo/integration/sessions_test.go:466]:
```suggestion
			if setSize < len(ops) {
```

File: x/mongo/driver/operation.go [hunk 5]
```diff
@@ -400,6 +400,16 @@ func (op Operation) Execute(ctx context.Context, scratch []byte) error {
 				return err
 			}
 			defer conn.Close()
+
+			// Set the server if the "session.Client" instance deferred the implicit session checkout.
```
Comment (benjirewis) [inline x/mongo/driver/operation.go:404]:
```suggestion
			// Set the server if it has not already been set and the session type is implicit.
```
The `if` check here is not _really_ checking if the `session.Client` instance deferred the implicit session checkout.

File: x/mongo/driver/session/client_session.go [hunk 6]
```diff
@@ -215,16 +215,31 @@ func NewClientSession(pool *Pool, clientID uuid.UUID, sessionType Type, opts ...
...
 
-	c.Server = servSess
-
 	return c, nil
 }
 
+// NewClientImplicitSessionDeferred will create a new implicit session client, defering the server checkout as a
+// user-defined operation.
+func NewClientImplicitSessionDeferred(pool *Pool, clientID uuid.UUID, opts ...*ClientOptions) (*Client, error) {
+
```
Comment (benjirewis) [inline x/mongo/driver/session/client_session.go:230]:
```suggestion
```

File: x/mongo/driver/session/client_session.go [hunk 7]
```diff
@@ -215,16 +215,31 @@ func NewClientSession(pool *Pool, clientID uuid.UUID, sessionType Type, opts ...
 		return nil, errors.New("causal consistency and snapshot cannot both be set for a session")
 	}
 
-	servSess, err := pool.GetSession()
-	if err != nil {
-		return nil, err
+	if sessionType != Implicit || !mergedOpts.deferImplicit {
```
Comment (benjirewis) [inline x/mongo/driver/session/client_session.go:218]:
```suggestion
	if sessionType != Implicit {
```
I'm a little confused why we need the new `deferImplicit` field on `session.ClientOptions`. Shouldn't we just only `SetServer` if it's an explicit session here, and otherwise wait until `operation.Execute` to call `SetServer` (after connection checkout).

File: x/mongo/driver/session/client_session.go [hunk 8]
```diff
@@ -215,16 +215,31 @@ func NewClientSession(pool *Pool, clientID uuid.UUID, sessionType Type, opts ...
...
 	}
 
-	c.Server = servSess
-
 	return c, nil
 }
 
+// NewClientImplicitSessionDeferred will create a new implicit session client, defering the server checkout as a
+// user-defined operation.
+func NewClientImplicitSessionDeferred(pool *Pool, clientID uuid.UUID, opts ...*ClientOptions) (*Client, error) {
```
Comment (benjirewis) [inline x/mongo/driver/session/client_session.go:229]: If we get rid of `deferImplicit`, I don't think this new constructor is necessary.

**Review by benjirewis — CHANGES_REQUESTED** [review]
Thanks so much for super helpful sequence diagram (what did you use to make that?)! 🧑‍🔧 Great test design, too. 

I have some questions about the general design and some smaller comments.

File: mongo/integration/sessions_test.go [hunk 9]
```diff
@@ -358,6 +360,121 @@ func TestSessions(t *testing.T) {
...
+				limitedSessionUse = true
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/integration/sessions_test.go:467]: Good catch, thank you!

File: x/mongo/driver/session/client_session.go (hunk 7 above)
Comment (prestonvasquez) [reply x/mongo/driver/session/client_session.go:218]: I think we want to default the behavior of `NewClientSession` to create an implicit session, as before. This ensures that existing functionality is preserved. The additional non-exported option allows us to extend the logic, however, wrapping this non-default case as the default behavior for `NewImplicitClientSessionDeferred`.

File: x/mongo/driver/session/client_session.go (hunk 8 above)
Comment (prestonvasquez) [reply x/mongo/driver/session/client_session.go:229]: See [this comment](https://github.com/mongodb/mongo-go-driver/pull/1027#discussion_r929121724). I went back and forth with this a lot, IMO it's better practice to preserve the existing functionality of `NewClientSession`. It's used in quite a few places and still used in some of the collection CRUD methods, such as `drop`.

File: mongo/integration/sessions_test.go (hunk 2 above)
Comment (prestonvasquez) [reply mongo/integration/sessions_test.go]: Add DRIVERS-2398 , which seems to fix the problem for our tests. Thank you @matthewdale !

File: mongo/integration/sessions_test.go [hunk 10]
```diff
@@ -358,6 +360,125 @@ func TestSessions(t *testing.T) {
...
+				minimumSessionCount = setSize
+			}
+
+			if setSize >= len(ops) {
+				limitedSessionUse = false
+			}
+		}
+
+		oneSessMsg := "expected one session across all %v operations for at least 1/%v retries, got: %v"
+		assert.True(mt, maintainedOneSession, oneSessMsg, len(ops), retrycount, minimumSessionCount)
```
Comment (matthewdale) [inline mongo/integration/sessions_test.go:478]: The `maintainedOneSession` assertion variable can be set to `true` on one loop and not reset back to `false` if a subsequent loop doesn't satisfy the `setSize == 1` condition. My understanding of the prose test is that we should make this assertion for every retry of the test logic. Move this assertion logic into the "retries" loop so we're asserting it on every retry.

E.g.
```go
// ...
setSize := len(set)
assert.True(mt,
	setSize == 1,
	"expected one session across all %v operations for at least 1/%v retries, got: %v",
	len(ops),
	retryCount,
	setSize)
// ...
```

File: mongo/integration/sessions_test.go [hunk 11]
```diff
@@ -358,6 +360,125 @@ func TestSessions(t *testing.T) {
 		assert.Equal(mt, err, mongo.ErrUnacknowledgedWrite,
 			"expected ErrUnacknowledgedWrite on unacknowledged write in session, got %v", err)
 	})
+	sessallocopts := mtest.NewOptions().ClientOptions(options.Client().SetMaxPoolSize(1).SetRetryWrites(true))
+	mt.RunOpts("14. implicit session allocation", sessallocopts, func(mt *mtest.T) {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/integration/sessions_test.go:366]: Hmm maybe? `TestSession` is where all of the existing prose tests are currently.

File: x/mongo/driver/session/client_session.go (hunk 7 above)
Comment (prestonvasquez) [reply x/mongo/driver/session/client_session.go:218]: Okay, sounds good.

Comment (prestonvasquez) [reply x/mongo/driver/session/client_session.go:218]: @benjirewis @matthewdale This change causes some failures in legacy write concern tests: https://evergreen.mongodb.com/build/mongo_go_driver_tests_42_plus_zlib_zstd_support__version~6.0_os_ssl_40~ubuntu1804_64_go_1_17_patch_f87317a97462447361004797b2e6d1023c61eaf4_62e008d13066152bc52df78e_22_07_26_15_31_48

Will have to investigate exactly why, but I'm sure there are collection methods, such as drop and aggregate, that can't defer implicit sessions in this exact way.

File: mongo/integration/sessions_test.go (hunk 10 above)
Comment (prestonvasquez) [reply mongo/integration/sessions_test.go:478]: The phrasing for [this](https://github.com/mongodb/specifications/blob/master/source/sessions/tests/README.rst#implicit-sessions-only-allocate-their-server-session-after-a-successful-connection-checkout) is kind of strange, here is the assertion:

> Assert the following across at least 5 retries of the above test:
Drivers MUST assert that exactly one session is used for all operations at least once across the retries of this test.

I interpret this to mean that of the five sessions that run, only one has to maintain only one session.

File: x/mongo/driver/operation.go [hunk 12]
```diff
@@ -400,6 +400,16 @@ func (op Operation) Execute(ctx context.Context, scratch []byte) error {
 				return err
 			}
 			defer conn.Close()
+
+			// Set the server if it has not already been set and the session type is implicit.
+			if op.Client != nil && op.Client.Server == nil && op.Client.SessionType == session.Implicit {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply x/mongo/driver/operation.go:405]: I don't have a strong opinion on this, but from my perspective it seems a little risky.

File: x/mongo/driver/session/client_session.go (hunk 7 above)
Comment (prestonvasquez) [reply x/mongo/driver/session/client_session.go]: This issue should be resolved by [this](https://github.com/mongodb/mongo-go-driver/pull/1027/files#diff-167dc8480d4039f98a2750ef31b20aac979efbb1c83466ac18cbf12c96b72b94R526-R533) block.

File: mongo/integration/mtest/mongotest.go [hunk 13]
```diff
@@ -521,7 +522,19 @@ func (t *T) ClearCollections() {
...
 				DropEncryptedCollection(t, coll.created, coll.CreateOpts.EncryptedFields)
 			}
-			_ = coll.created.Drop(context.Background())
+
+			c := coll.created
+			err := c.Drop(context.Background())
+			if err == mongo.ErrUnacknowledgedWrite || err == driver.ErrUnacknowledgedWrite {
+				// It's possible that a collection could have an unacknowledged write concern, which
+				// could prevent it from being dropped for sharded clusters. We can resolve this by
+				// re-instantiating  the collection with a majority write concern before dropping.
```
Comment (benjirewis) [inline mongo/integration/mtest/mongotest.go:531]:
```suggestion
				// re-instantiating the collection with a majority write concern before dropping.
```
Good find! 🧑‍🔧

File: mongo/integration/mtest/mongotest.go [hunk 14]
```diff
@@ -521,7 +522,19 @@ func (t *T) ClearCollections() {
 			if coll.CreateOpts != nil && coll.CreateOpts.EncryptedFields != nil {
 				DropEncryptedCollection(t, coll.created, coll.CreateOpts.EncryptedFields)
 			}
-			_ = coll.created.Drop(context.Background())
+
+			c := coll.created
+			err := c.Drop(context.Background())
```
Comment (benjirewis) [inline mongo/integration/mtest/mongotest.go:527]: Can we just do `err := coll.created.Drop(context.Background())` and then use a new variable on L534?

File: mongo/integration/sessions_test.go [hunk 15]
```diff
@@ -358,6 +360,121 @@ func TestSessions(t *testing.T) {
...
+				cmd := cmd
+				errs.Go(func() error {
+					if err := op(ctx); err != nil {
+						return fmt.Errorf("error running %s operation: %v", cmd, err)
+					}
+					return nil
+				})
+			}
+			err := errs.Wait()
+			assert.Nil(mt, err, "expected no error, go: %v", err)
```
Comment (benjirewis) [inline mongo/integration/sessions_test.go:449]:
```suggestion
			assert.Nil(mt, err, "expected no error, got %v", err)
```

File: mongo/integration/sessions_test.go (hunk 10 above)
Comment (benjirewis) [reply mongo/integration/sessions_test.go:472]: @prestonvasquez that is my interpretation as well and I believe your current assertion is correct.

File: x/mongo/driver/operation.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (benjirewis) [reply x/mongo/driver/operation.go:407]: I would rather leave the explicit session logic as-is if it's not required to be moved.

File: x/mongo/driver/session/client_session.go [hunk 16]
```diff
@@ -215,16 +215,22 @@ func NewClientSession(pool *Pool, clientID uuid.UUID, sessionType Type, opts ...
 		return nil, errors.New("causal consistency and snapshot cannot both be set for a session")
 	}
 
-	servSess, err := pool.GetSession()
-	if err != nil {
-		return nil, err
+	if sessionType == Explicit {
```
Comment (benjirewis) [inline x/mongo/driver/session/client_session.go:218]: Can we leave a comment above this explaining why we only `SetServer` here for `Explicit` sessions and `Implicit` sessions call `SetServer` after connection checkout in `Execute`?

File: mongo/integration/mtest/mongotest.go (hunk 14 above)
Comment (prestonvasquez) [reply mongo/integration/mtest/mongotest.go:527]: Yep, that's fine. This was because I was initially asserting that an error doesn't happen but that's quite a different problem.

File: x/mongo/driver/session/client_session.go (hunk 16 above)
Comment (prestonvasquez) [reply x/mongo/driver/session/client_session.go:218]: Absolutely.

File: mongo/integration/mtest/mongotest.go (hunk 13 above)
Comment (benjirewis) [reply mongo/integration/mtest/mongotest.go:530]: Bump.

**Review by benjirewis — APPROVED** [review]
Looks great! 🧑‍🔧

File: mongo/integration/sessions_test.go (hunk 10 above)
Comment (matthewdale) [reply mongo/integration/sessions_test.go:472]: Agreed, I reached the same conclusion after re-reading it.

File: x/mongo/driver/operation.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply x/mongo/driver/operation.go:407]: I don't have a strong opinion either way.

**Review by matthewdale — APPROVED** [review]
Looks good 👍
